### DISCUSSION TITLE
Using PointSkyRegion for simulations of drifting instruments?

### DISCUSSION THREAD

**Discussion by mdpunch** [gammapy/gammapy#4069] [Help]
Hi all (especially those working on drifting instruments),

(using gammapy 0.20.1)

I'm trying to adapt to using `PointSkyRegion` for the analysis of ALTO simulations (currently using `CircleSkyRegion`, with workarounds).

The IRFs are produced for a fixed Zenith angle (18degrees for gammas, 15-21degrees for background), and with variable `RAD_MAX` with energy.

So, the "pointing" is Zenith ```pointing = SkyCoord("0 deg", "90 deg")``` and the `Observation` is created with this and the IRFs.  The source region is then 18deg from Zenith ```center = SkyCoord("0 deg", "72 deg")```, so ```on_region = PointSkyRegion(center=center)```.

Then the usual `SpectrumDataset` creation:
```
geom = RegionGeom.create(region=on_region, axes=[energy_axis])
empty_dataset = SpectrumDataset.create(geom=geom, energy_axis_true=energy_axis_true)
spectrum_maker = SpectrumDatasetMaker(
    containment_correction=False, selection=[#"counts", 
        "exposure", "edisp","background"]
)
dataset = spectrum_maker.run(empty_dataset, obs)
```
and for the background regions:
```
region_finder = WobbleRegionsFinder(n_off_regions=12)
bkg_maker = ReflectedRegionsBackgroundMaker(region_finder=region_finder)
```

But, when trying to make the `dataset_on_off`, the background maker doesn't seem to like having an Alt/Az pointing:
```
dataset_on_off = bkg_maker.run(dataset, observation=obs)

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_1846566/3457672469.py in <cell line: 1>()
----> 1 dataset_on_off = bkg_maker.run(dataset, observation=obs)

/scr/punch/miniconda3/envs/alto/lib/python3.9/site-packages/gammapy/makers/background/reflected.py in run(self, dataset, observation)
    559             On off dataset.
    560         """
--> 561         counts_off, acceptance_off = self.make_counts_off(dataset, observation)
    562         acceptance = RegionNDMap.from_geom(geom=dataset.counts.geom, data=1)
    563 

/scr/punch/miniconda3/envs/alto/lib/python3.9/site-packages/gammapy/makers/background/reflected.py in make_counts_off(self, dataset, observation)
    513 
    514         if geom.is_all_point_sky_regions and len(regions_off) > 0:
--> 515             regions_off = self._filter_regions_off_rad_max(
    516                 regions_off, energy_axis, geom, events, observation.rad_max
    517             )

/scr/punch/miniconda3/envs/alto/lib/python3.9/site-packages/gammapy/makers/background/reflected.py in _filter_regions_off_rad_max(regions_off, energy_axis, geom, events, rad_max)
    452     def _filter_regions_off_rad_max(regions_off, energy_axis, geom, events, rad_max):
    453         # check for overlap
--> 454         offset = geom.region.center.separation(events.pointing_radec)
    455 
    456         e_min, e_max = energy_axis.bounds

AttributeError: 'NoneType' object has no attribute 'pointing_radec'
```

Has anyone succeeded in using the `PointSkyRegion` this way, or is there another kind of workaround in order to define the background regions?

Thanks in advance for corrections/suggestions,

Michael.

**Comment by adonath** [comment]
Thanks @mdpunch! The error message suggests, that the error is related to the `Observation` object not having any `events`. Which I guess is expected, because it is a simulation? For the background estimation however you would need those, otherwise there is nothing you could estimate the background from. There are two options:
- Simulate an event list (which requires a 3D background model) and continue using your proposed workflow. See https://docs.gammapy.org/0.20.1/tutorials/analysis/3D/event_sampling.html
- Assume some kind of background spectrum and simulate the off data from that, see https://docs.gammapy.org/0.20.1/tutorials/analysis/1D/spectrum_simulation.html

I hope this helps.

**Reply by mdpunch** [reply]
Hi @adonath,

That could give me some clues, so I'll look into those.

One thing I notice immediately as a difference between `PointSkyRegion` and `CircleSkyRegion` is that for the latter, I already have background counts after I run the `spectrum_maker`. But then I have to adjust the background counts as a function of energy in order to take the energy-dependent angular cut into account.

Is there some deep reason why there would be background counts after running the `spectrum_maker` for the `CircleSkyRegion`, but not for the `PointSkyRegion`?  That might also give me some ideas...

Thanks,
Michael.

**Reply by adonath** [reply]
The behavior you observe does not make sense to me. The dataset can only contain background counts after running the `spectrum_maker` if the `Observation` object contains events. If it contains events the `PointSkyRegion` should work equally well.
How do you create your observation object?

**Reply by mdpunch** [reply]
Hi Axel,

So, for the details when I use `CircleSkyRegion`:

```
pointing = SkyCoord("0 deg", "90 deg")
obs = Observation.create(pointing=pointing, irfs=irfs, livetime=livetime)

center = SkyCoord("0 deg", "72 deg")

# Create on_region (CircleSkyRegion)
# The value of test_rad doesn't matter, as long as it is the same in the region(_on)
# ... and in the calculation of the factor to adjust the off by (dependent on the theta2 cut)
test_rad = 1.00 * u.deg
on_region = CircleSkyRegion(center=center, radius=test_rad)

# Make empty Dataset
geom = RegionGeom.create(region=on_region, axes=[energy_axis])
empty_dataset = SpectrumDataset.create(geom=geom, energy_axis_true=energy_axis_true)

# Make Dataset (with CircleSkyRegion)
spectrum_maker = SpectrumDatasetMaker( selection=["exposure", "edisp", "background"],)
dataset = spectrum_maker.run(empty_dataset, obs)
```
at the end of this, the dataset looks like:
```
SpectrumDataset
---------------

  Name                            : BCIPooek 

  Total counts                    : 0 
  Total background counts         : 226704.27
  Total excess counts             : -226704.27

  Predicted counts                : 226704.27
  Predicted background counts     : 226704.27
  Predicted excess counts         : nan

  Exposure min                    : 3.69e+05 m2 s
  Exposure max                    : 1.57e+10 m2 s

  Number of total bins            : 20 
  Number of fit bins              : 20 

  Fit statistic type              : cash
  Fit statistic value (-2 log(L)) : nan

  Number of models                : 0 
  Number of parameters            : 0
  Number of free parameters       : 0
```
... which has the expected number of background counts.

Whereas with the PointSkyRegion, the dataset has 0 background counts at this point... so maybe that could give you a clue to the differences?

**Reply by adonath** [reply]
Thanks @mdpunch, this was helpful. My misunderstanding was that I though the background was derived from the data, as an "OFF" observation. However this is not the case here. In your example the background is computed from a "background IRF". This requires integrating the background model over the region size, which is zero for the point sky region, that's why the resulting background rate is zero as well. Currently the energy dependent on region size is only support for the case, where the background is estimated from the data. 

However you can work-around the issue, e.g. like so:

```
offset = 
energy = 

rad = obs.rad_max(offset=offset, energy=energy)
solid_angle =  2 * np.pi * (1 - np.cos(rad)) 
bkg_rate = obs.background.evaluate(offset=offset, energy=energy)
dataset.background.data = bkg_rate * obs.observation_livetime * solid_angle 
```
The code is not guaranteed to work, but you get the idea...I hope that helps.

**Reply by mdpunch** [reply]
Thanks for the suggestion, Axel !

It does need a bit of work on my side, since the max_rad binning is different to the finer bkg binning, so it looks something like this:

```
rebin = bkg.axes['energy'].nbin//energy_axis.nbin

tht_cut = obs.rad_max.evaluate(offset=18*u.deg,energy=energy_axis.center)
tht_cut[tht_cut == 0] = -1e-2*u.deg         # !!! HACK replace 0's by -1e-2deg, since otherwise gives errors later

sol_ang = 2*np.pi*(1-np.cos(tht_cut)) *u.sr
bkg_rate = bkg.data.T[0] * bkg.unit * bkg.axes['energy'].bin_width.to(u.MeV)
bkg_rate_rebin = np.sum(bkg_rate.reshape(-1,rebin),axis=1)
counts_off_new = (bkg_rate_rebin * obs.observation_live_time_duration * sol_ang)
dataset.background.data = (counts_off_new).value
```
... then later:
```
acceptance_on = RegionNDMap.from_geom(dataset.counts.geom,data=sol_ang.value,unit=sol_ang.unit)
acceptance_off = RegionNDMap.from_geom(dataset.counts.geom,data=bkg_sim_sr.value,unit=bkg_sim_sr.unit)

# Trying with both acceptance_on and acceptance_off being Maps, seems to work
dataset_on_off = SpectrumDatasetOnOff.from_spectrum_dataset(
    dataset=dataset, 
    acceptance= acceptance_on,  
    acceptance_off = acceptance_off,
)
```
... give similar results to my CircleSkyRegion hack, and I guess is more correct (will check in more detail now, and post if there are big discrepencies).

**Comment by mdpunch** [comment]
Thanks Axel @adonath, I can confirm that with my modified version of your suggested code for the `PointSkyRegion`, I get about the same result as my `CircleSkyRegion` hack (just a tiny bit worse), and with cleaner code (no assumed `test_radius` for the `CircleSkyRegion` which has to be corrected for later).

**Comment by adonath** [comment]
@registerrier Do you think we should forbid the use of the `PointSkyRegion` for the `SpectrumDatasetMaker` with background templates and energy dependent rad max? Or make it work instead?

**Reply by registerrier** [reply]
In a regular analysis you probably would not have `beackground` IRFs attached to data in `point-like` mode.  So it should fail already. 

Now, for simulations the situation is different because you would need to have a `background` IRF to estimate background counts in the energy dependent ON region. So this probably means we need to support energy dependent rad max background estimations at some point. 

I suppose that we should make the `SpectrumDatasetMaker` properly raise an error for background estimation with `PointSkyRegion` for now and create an issue to make this work later.

**Reply by mdpunch** [reply]
Hi Regis, Axel,

But it's working for me now, with Axel's suggestion modified!

So, I would prefer that it continue to do so, with `SpectrumDatasetMaker`.

Maybe have the error rather at the `SpectrumDatasetOnOff.from_spectrum_dataset` level, to raise an error only if the `dataset.background` is zero, and to leave to run it if someone (like me) has calculated the background from the IRF to put in there?

Actually, in the current state of things the `SpectrumDatasetOnOff.from_spectrum_dataset` already gives an pretty nonsense result with a zero `dataset.background`, as I just checked:
```
SpectrumDatasetOnOff
--------------------

  Name                            : rQ3ey7BU 

  Total counts                    : 0 
  Total background counts         : nan
  Total excess counts             : nan

  Predicted counts                : nan
  Predicted background counts     : 0.00
  Predicted excess counts         : nan
```
So, maybe just raise an error if there are NaN at this point?

Michael.

**Reply by mdpunch** [reply]
P.S. with my/Axel's modification, the spectrum dataset looks like this
```SpectrumDataset
---------------

  Name                            : Efzlajyh 

  Total counts                    : 0 
  Total background counts         : 147725.78
  Total excess counts             : -147725.78

  Predicted counts                : 147725.78
  Predicted background counts     : 147725.78
  Predicted excess counts         : nan

  Exposure min                    : 3.69e+05 m2 s
  Exposure max                    : 1.57e+10 m2 s

  Number of total bins            : 20 
  Number of fit bins              : 20 

  Fit statistic type              : cash
  Fit statistic value (-2 log(L)) : nan

  Number of models                : 0 
  Number of parameters            : 0
  Number of free parameters       : 0
```
and the spectrumOnOff dataset:
```
SpectrumDatasetOnOff
--------------------

  Name                            : a9jVNaIe 

  Total counts                    : 0 
  Total background counts         : 147725.78
  Total excess counts             : -147725.78

  Predicted counts                : 146721.39
  Predicted background counts     : 146721.39
  Predicted excess counts         : nan

  Exposure min                    : 3.69e+05 m2 s
  Exposure max                    : 1.57e+10 m2 s

  Number of total bins            : 20 
  Number of fit bins              : 20 

  Fit statistic type              : wstat
  Fit statistic value (-2 log(L)) : 294444.53

  Number of models                : 0 
  Number of parameters            : 0
  Number of free parameters       : 0

  Total counts_off                : 27267290 
  Acceptance                      : 20 
  Acceptance off                  : 6509831 
```
with `dataset.background.plot()`
![image](https://user-images.githubusercontent.com/17853426/189720101-4fb0f3a1-5e3e-427f-a204-b708b29c6181.png)
